### ISSUE TITLE
Doc: [no-shadow] Specialized error on enum type with the same name as one of the values

### ISSUE THREAD

**Issue by artificialWave** [typescript-eslint/typescript-eslint#9755]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=KYOwrgtgBAogHgQwgBwDbCgbwLACgpQByA9gC7xJrAA0eBFK6tuAvnkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaZACwEMATfAd3UnwDMHIAacbSAAQBcBPABxQBjaAEt%2B3Qinijc3APQFi5KrTSRE0aPmiR2AXxD6gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
enum Example {
  NotExample,
  Example,
}
```

### ESLint Config

```javascript
{
  "rules": {
    "no-shadow":"off",
    "@typescript-eslint/no-shadow":"error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect that this works, and does not result in any linting errors. At least it used to work before upgrading to `@typescript-eslint` v8.

### Actual Result

`'Example' is already declared in the upper scope on line 1 column 6. 3:3 - 3:10`

**Comment by kirkwaiblinger** [comment]
@bradzacher 

Question - is this not expected per the docs? https://typescript-eslint.io/rules/no-shadow#why-does-the-rule-report-on-enum-members-that-share-the-same-name-as-a-variable-in-a-parent-scope

**Comment by bradzacher** [comment]
Yes - this is working as intended.

Enums can be self-referenced within their body.
And enum members can be referenced without a namespace within their body.

For example:
```ts
enum Foo {
  A,
  B = A,
}

enum Bar {
  C,
  D = Bar.C,
}
```

So by defining a member with the same name as the function you've shadowed the enum name and made it unreachable.

This is conceptually the same as declaring a variable with the same name as the function you're in, which also reports in the rule.

```ts
function foo() {
  let foo = 1;
  return foo;
}
```
[playground](https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=KYOwrgtgBAogHgQwgBwDbCgbwLACgpQByA9gC7xJrAA0eBFK6tuAvnngGZggDGpAlsRBQOxYgAoAlFjpR0pEWKgBeKAEYA3LIBOwUmG3DRxLayA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaZACwEMATfAd3UnwDMHIAacbSAAQBcBPABxQBjaAEt%2B3Qinijc3APQFi5KrTSRE0aPmiR2AXxD6gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

So yeah we can close this - it's fully working as expected, I think.

**Comment by JoshuaKGoldberg** [comment]
Should we add a specialized error message?

**Comment by Josh-Cena** [comment]
I agree we should have a specialized error message reminding people that enum values are added to the enum scope, because we have received so many issues about `[declaration of X]; enum Foo { X }` not working.

**Comment by bradzacher** [comment]
We could - but by-and-large the reports have stopped since we added the FAQ to the rule docs.

This is a new report in v8 though cos we changed the AST and scope structure which allowed us to detect it.

We could update the docs but I doubt it's a case many people would see. It's quite rare to name a member the same as the enum itself.

**Comment by Josh-Cena** [comment]
I'm not referring specifically to the case of enum member shadowing enum name, just in general when enum values shadow upper declarations.

**Comment by JoshuaKGoldberg** [comment]
I'm +1 on adding that general error specialization. Given that the docs FAQ has been proven useful, I'd wager getting it in the editor is the same usefulness, faster.

**Comment by bschlenk** [comment]
Would it make sense to only report this as an issue if you're assigning values to the enum members? If you're omitting the values and relying on typescript assigning incrementing numbers, then there is no shadowing going on.

**Comment by kirkwaiblinger** [comment]
@bschlenk 

> Would it make sense to only report this as an issue if you're assigning values to the enum members? If you're omitting the values and relying on typescript assigning incrementing numbers, then there is no shadowing going on.

This conversation came up before but it seems like we didn't have consensus to change anything about the behavior: https://github.com/typescript-eslint/typescript-eslint/issues/10749#issuecomment-2626246106

**Comment by bradzacher** [comment]
My 2c is that no it shouldn't skip them.
Just like the rule doesn't skip this code:

```ts
const foo = 1;
function fn() {
  const foo = 2;
}
```

Usage of a declaration does not change the fact that it is shadowing something.
